include/deal.II/base/work_stream.h

```diff
@@ -874,14 +1019,352 @@ namespace WorkStream
+                                                sample_copy_data);
 
+              parallel::internal::parallel_for(
+                colored_iterators[color].begin(),
+                colored_iterators[color].end(),
+                [&worker_and_copier](
+                  const tbb::blocked_range<
+                    typename std::vector<Iterator>::const_iterator> &range) {
+                  worker_and_copier(range);
+                },
+                chunk_size);
+            }
+      }
 
+    }    // namespace tbb_colored
 #  endif // DEAL_II_WITH_TBB
 
 
+
+#  ifdef DEAL_II_WITH_TASKFLOW
+    namespace taskflow_v4
+    {
+      template <typename CopyData>
+      class Chunk
+      {
+      public:
+        Chunk(const unsigned int count, const CopyData &copy_data)
+          : copy_datas(count, copy_data)
+        {}
+        std::vector<CopyData> copy_datas;
+      };
+
+
+
+      /**
+       * WorkStream without colored iterators.
+       *
+       * Work in chunks of size @p chunk_size (only if we have enough items, otherwise 1) and
+       * use a thread-local scratch object
+       */
+      template <typename Worker,
+                typename Copier,
+                typename Iterator,
+                typename ScratchData,
+                typename CopyData>
+      void
+      run(const Iterator &                         begin,
+          const typename identity<Iterator>::type &end,
+          Worker                                   worker,
+          Copier                                   copier,
+          const ScratchData &                      sample_scratch_data,
+          const CopyData &                         sample_copy_data,
+          const unsigned int                       chunk_size = 8)
+      {
+        if (MultithreadInfo::n_threads() == 1)
+          {
+            // need to copy the sample since it is marked const
+            ScratchData scratch_data = sample_scratch_data;
+            CopyData    copy_data    = sample_copy_data; // NOLINT
+
+            for (Iterator i = begin; i != end; ++i)
+              {
+                // need to check if the function is not the zero function. To
+                // check zero-ness, create a C++ function out of it and check
+                // that
+                if (static_cast<const std::function<void(
+                      const Iterator &, ScratchData &, CopyData &)> &>(worker))
+                  worker(i, scratch_data, copy_data);
+                if (static_cast<const std::function<void(const CopyData &)> &>(
+                      copier))
+                  copier(copy_data);
+              }
+
+            return;
+          }
+
+        tf::Executor &executor = MultithreadInfo::get_taskflow_executor();
+        tf::Taskflow  taskflow;
+
+        tf::Task last_copier = taskflow.placeholder();
+
+        Threads::ThreadLocalStorage<std::unique_ptr<ScratchData>>
+          thread_local_scratch;
```

Comment:
This does not work. It runs into the same bug we took so many hours to figure out back when. You need to have a list of scratch datas for every thread.